### DISCUSSION TITLE
Request for help

### DISCUSSION THREAD

**Discussion by martin-san03** [offa/influxdb-cxx#135] [Q&A] [answered]
Hello, sorry I'm quite new in C++, this is surelly an stupid question.
I have an array of measurements with different number of fields. I'm not sure what is the better way (no-memcopy) to add multiple fields in a for loop to a influxdb::Point and write that point. 

The following pattern is not possible for me:
`influx->write(influx::point{"name"}.addField("fieldname1", value1).addField("fieldname2", value2) .... ;`

And I soon as I start doing things like:

```
1.    // const std::vector<const char*>* fields is pointer to std::vector<const char*, std:vector<const * char>
2.    influxdb::Point && point = influxdb::Point{ measurementName };
3.
4.    for (int j = 0; j<(*fields).size(); j++)
5.    {
6.        const char* fieldName = (*fields)[i];
7.        point.addField(fieldName, std::string("TODO: add value"));
8.    }
9.
10.   influx->write(point);
```
                
I get the following errors:
```
E0413: no suitable conversion function from "std::string" to "int" exists in line 7
E0304: no instance of overloaded function "influxdb::InfluxDB::write" matches the argument in line 10
```

A little help would be appreciated.

**Comment by martin-san03** [comment]
Seems I was compiling C++ 14....I have resolved all the compile errors except for 

`10.  influx->write(point);`

**Comment by martin-san03** [comment]
So, I workaround the problem by adding a tag to the write statement:

`10. influx->write(point.addTag("dummy","dummy");`

resolved some linking errors and finally got the app running. Now I'm getting an BadRequest exception in HTTP::treatCurlResponse with response CURLE_OK and responseCode 401.

I think this is because authentification problems.

I added a user InfluxStreamer with all privileges and set it in the connection string:

`http://InfluxStreamer:InfluxStreamer123@localhost:8086?db=db_name`

and I'm able to connect and see the data using the web client at http://localhost:8086, putting the credentials when asked to log in in

`http://localhost:8086/signin`

but I'm still not able to connect through my application. 
when I try to use the command line client I have also the same request:

```
PS C:\influxDb\client-2.4.0> .\influx.exe  bucket list --org org_name
Error: failed to list buckets: 401 Unauthorized: unauthorized access
```

I'm using version 2.4.0 on windows, as you may see in the previous command.
I have read that I should use Tokens, but I have also found they are not supported yet. Is it true?

Many thanks in advance

**Comment by martin-san03** [comment, marked as answer]
I was missing a auth v1 grant in the database. Now it is functionning correctly, excepting for one thing. The value show in grafana or influxdb UI is incorrect, it seems influxdb uses a reference to the value provided to Point.addField, but the variable was a temporal one.
This is my code:

```
    // convert bytes in memory into appropiate type
    static bool addValue(influxdb::Point* point, const char * fieldName, const char* typeName, const char* addr)
    {
        bool result = true;

        if (::strcmp(typeName, "int8_t") == 0)
            point->addField(fieldName, (int)*reinterpret_cast<const int8_t*>(addr));
        else if (::strcmp(typeName, "int16_t") == 0)
            point->addField(fieldName, (int)*reinterpret_cast<const int16_t*>(addr));
        else if (::strcmp(typeName, "int32_t") == 0)
            point->addField(fieldName, (int)*reinterpret_cast<const int32_t*>(addr));
        ...
    }

    ....
   bool server()
   {
        ....
        auto influx = influxdb::InfluxDBFactory::Get(pdi::influxURL);
        ....
        // iterates all the configured measurands, and send each measurand field of them
        for (int i = 0; i < measurements.size(); i++)
        {
            const std::tuple< const mw::Node*, const std::vector< const mw::Node*>> measurement = measurements[i];
            const mw::Node* measurementNode = std::get<0>(measurement);
            const std::vector< const mw::Node*> * fields = &std::get<1>(measurement);

            influxdb::Point && point = influxdb::Point{ measurementNode->name };

            for (int j = 0; j<(*fields).size(); j++)
            {
                const mw::Node* fieldNode = (*fields)[j];
                assert(addValue(&point, fieldNode->name, fieldNode->typeName, fieldNode->memAddr));
            }

            influx->write(point.addTag("hostName", pdi::hostName));
        }

        // be sure data has been sent (normally already sent in last write)
        influx->flushBatch();
```
it seems the value passed to the point at line:
`point->addField(fieldName, (int)*reinterpret_cast<const int32_t*>(addr));`

is garbage when `influxdb->write()` or `influxdb->flushBatch()` is called.
If I do as suggested in tutorials, the values are correct in influxdb UI:

```
influxdb->write(Point{"measurementName"}.addField("field1",1).addField("field2",2).addTag("tagname","tagvalue");
influxdb->flushBatch();
```

Any help is really appreciated....I will continue investigating on my own anyway.

**Reply by offa** [reply]
First, you should try to avoid C code. Using `strcmp` and `reinterpret_cast` are likely wrong. :-)

It's hard to understand what happens here. So first thing to do is: Ensure your `addField()` function yields a valid field value.

```
influxdb::Point && point = influxdb::Point{ measurementNode->name };
[…]
assert(addValue(&point, fieldNode->name, fieldNode->typeName, fieldNode->memAddr));
```
This is likely wrong as `&point` takes an address of a R-value. I'd suggest avoiding pointers here anyway and return the actual result eg. as variant (as `addField()` takes as argument). You should also replace the C-String (`const char*`) with C++ APIs like `std::string` / `std::string_view`.

**Reply by offa** [reply]
Also it might become handy to disassemble the problem into smaller parts: (a) get valid values that can be written and (b) writing to database.

**Reply by martin-san03** [reply]
Thanks for the feedback. I will use C++ types instead. Concerning the problem and the understaing of the problem, I will try to provide more information:

We have an internal data structure that olds the last output of each software component. 

```
root
   |--components
                    |- name: string
                    |- output
                              |- name: string
                              |- typeName: string
                              |- value_address: void* (pointer to memory address of int, float, char, enum, ...)    
```

This strucutre is dynamically created at runtime. We want to stream to influxdb the status of the output of each software component.

component->name would be the measurement, and each of the output would be the fileds of a influxdb::Point.
output values can be of different types (int, string, enums, etc.). we hold the typename and a pointer to the address where the value is stored. So we need to convert the pointer to the address into the actual int, float, string variable before adding it as a field to the point.
This is what addValue() is doing. The reinterpret_casts are there to do the type conversion from a void* to the correct pointer type.
Once converted, the dereferenced pointer is added to the Point.

In server() function, we iterate all the components and its outputs and we add a point for each component, consisting of fields (each output) and a tag (the host name). This is the function that calls addValue(), to interpret the output type, cast the output value and add it to the point.

If I do a breakpoint just before the influx->write. I see the Point contains correct data, the measurement name, all the fields and the tags are there, with correct names and values. But this data is not the one influxdb UI shows.

if instead of all this, I do:

```
int output1_value = 0;
float output2_value= 1.0;
influx->write(Point{"component"}.addField("Output1", output_1value).addField("Output2", output2_value);
influx->flushBatch();
```

these values are correctly observed in influxdb UI.
So I'm doing something wrong in the manipulation of point. My hypothesis is that this conversions in addValue() leads to a local temporal memory that do not exists when the actual data is sent to the influxdb (influx->write or influx->flushBatch())
My limited knownledge on C++ does not allow me to understand what is happening :(

**Reply by martin-san03** [reply]
and I will try to return the a std::variants from addValue() as you suggest

**Reply by martin-san03** [reply]
Ok, works now.....I cannot beleave it. This was the problem:

![imagen](https://user-images.githubusercontent.com/115890440/196236101-12bc5a03-65e9-4a8e-a46b-47d0b2fdf782.png)

mean was selected :( creating a wired pattern:

![imagen](https://user-images.githubusercontent.com/115890440/196236322-e8f42270-e6c8-4c48-8337-3cf73e8f0596.png)

Sorry for the worries....Thanks a lot!

**Reply by offa** [reply]
Nice! :-)
